autoload/ale/code_action.vim

```diff
@@ -0,0 +1,73 @@
+function! ale#code_action#HandleCodeAction(code_action) abort
+    let l:current_buffer = bufnr('')
+    let l:existing_buffers = {}
+    let l:changes = a:code_action.changes
+
+    for l:file_code_edit in l:changes
+        let l:buf = ale#util#GetBufferNumber(l:file_code_edit.fileName)
+
+        if l:buf != -1
+            let l:existing_buffers[l:buf] = 1
+
+            if l:buf != l:current_buffer && ale#util#IsModified(l:buf)
+                call ale#util#Execute('echom ''Aborting action, file is unsaved''')
```

Comment:
I reckon just apply the changes anyway, even if buffers are modified. The changes will often be merged into files you're currently editing, and so they are likely to be modified. The chance of modifying something someone didn't want to save is rare. If we do want to not modify something for some reason, then don't change buffers for them, and don't apply any changes. Let them change buffers.

Reply:
I added this check because tsserver could return wrong ranges in case a buffer was modified, since it will read the files from the disk and not from the buffer. For example if somebody tries to rename the `value` symbol from `file2.ts`:

### file1.ts
```typescript
/* 1 */ class A {
/* 2 */   value: number
/* 3 */ }
```

### file2.ts
```typescript
/* 1 */ import {A} from './file1'
/* 2 */ const a = new A()
/* 3 */ console.log(a.value)
```

If `file1.ts` was modified and a line was added above line 2, then the result from `tsserver` would not match the actual contents of the buffer.

Reply:
Or are you saying the buffers should be saved before sending any requests to tsserver?